camel/data_collector/sharegpt_collector.py

```diff
@@ -0,0 +1,173 @@
+        ]
+    }
+"""
+
+
+class ConversationItem(BaseModel):
+    from_: Literal["human", "gpt", "function_call", "observation"]
+    value: str
+
+    class Config:
+        fields: ClassVar[Dict[str, str]] = {"from_": "from"}
+        extra = "forbid"
+
+
+class ShareGPTData(BaseModel):
+    system: str
+    tools: str
+    conversations: List[ConversationItem]
+
+    class Config:
+        extra = "forbid"
+
+
+class ShareGPTDataCollector(BaseDataCollector):
+    def __init__(self) -> None:
+        super().__init__()
+        self.system_message: Optional[BaseMessage] = None
+        self.agent_name: Optional[str] = None
+        self.tools: List[FunctionTool] = []
+
+    def inject(
+        self,
+        agent: Union[List[ChatAgent], ChatAgent],
+    ) -> Self:
+        r"""Inject an agent into the data collector."""
+        if not self.agent_name:
+            _agent = agent if isinstance(agent, ChatAgent) else agent[0]
+            self.agent_name = _agent.role_name
+            self.system_message = _agent._system_message
+            self.tools += list(_agent.tool_dict.values())
+
+        super().inject(agent)
+        return self
+
+    def convert(self) -> Dict[str, Any]:
+        r"""Convert the collected data into a dictionary."""
+        if self.agent_name is None:
+            raise ValueError("No agent injected")
+        if history := self.get_agent_history(self.agent_name):
+            data = dict(
+                system=self.system_message.content
+                if self.system_message
+                else "",
+                tools=json.dumps(
+                    [
+                        t.get_openai_tool_schema()["function"]
+                        for t in self.tools
+                    ]
+                ),
+                conversations=[],
+            )
+            conversations: List[Any] = []
+            for _data in history:
+                role, message = _data.role, _data.message
+                if role == OpenAIBackendRole.USER:
+                    conversations.append(
+                        {"from": "human", "value": message.content}
+                    )
+                elif role == OpenAIBackendRole.ASSISTANT:
+                    if isinstance(message, FunctionCallingMessage):
+                        tmp = dict(
+                            name=message.func_name,
+                            arguments=message.args,
+                        )
+                        conversations.append(
+                            {"from": "function_call", "value": json.dumps(tmp)}
+                        )
+                    else:
+                        conversations.append(
+                            {"from": "gpt", "value": message.content}
+                        )
+                elif role == OpenAIBackendRole.FUNCTION:
+                    conversations.append(
+                        {
+                            "from": "observation",
```

Comment:
Would be good to make the role here configurable. Also I think checking if the message has a result or calls is a more robust way of differentiating between function call and function result (and tool call and tool result in the future), until we have some better type safety in this area

Reply:
Hi Caelum, 

I've already switched to using memory to retrieve history and roles. Regarding "Would be good to make the role here configurable," I'm not entirely sure I fully understand what you mean. Could you clarify? Are you suggesting making the roles customizable in some way?